Added quality declaration (#90)
ros2/rmw_implementation: QUALITY_DECLARATION.md

This document is a declaration of software quality for the `rmw_implementation` package, based on the guidelines in [REP-2004](https://www.ros.org/reps/rep-2004.html).

# `rmw_implementation` Quality Declaration

The package `rmw_implementation` claims to be in the **Quality Level 4** category.

Below are the rationales, notes, and caveats for this claim, organized by each requirement listed in the [Package Requirements for Quality Level 1 in REP-2004](https://www.ros.org/reps/rep-2004.html).

## Version Policy [1]

### Version Scheme [1.i]

`rmw_implementation` uses `semver` according to the recommendation for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#versioning).

### Version Stability [1.ii]

`rmw_implementation` is not yet at a stable version, i.e. >= 1.0.0.

### Public API Declaration [1.iii]

**`rmw_implementation` does not expose a public API.**

> ### API Stability Within a Released ROS Distribution [1.iv]/[1.vi]
> 
> `rmw_implementation` will not break public API within a released ROS distribution, i.e. no major releases once the ROS distribution is released.
> 
> ### ABI Stability Within a Released ROS Distribution [1.v]/[1.vi]
> 
> `rmw_implementation` will maintain ABI stability within a ROS distribution.
> 

Reviewer: I think it's better leaving [v.i] on its own, even if it can be considered repeated information

```suggestion
### API Stability Policy [1.iv]

`rmw_implementation` will not break public API within a released ROS distribution, i.e. no major releases once the ROS distribution is released.

### ABI Stability Policy [1.v]

`rmw_implementation` contains C code and therefore must be concerned with ABI stability, and will maintain ABI stability within a ROS distribution.

### ABI and ABI Stability Within a Released ROS Distribution [1.vi]

`rmw_implementation` will not break API nor ABI within a released ROS distribution, i.e. no major releases once the ROS distribution is released.
```

### Copyright Statement [3.iv]

The copyright holders each provide a statement of copyright in each source code file in `rmw_implementation`.

There is an automated test which runs a linter (ament_copyright) that ensures each file has at least one copyright statement.

Most recent test results can be found [here](http://build.ros2.org/view/Epr/job/Epr__class_loader__ubuntu_bionic_amd64/lastBuild/testReport/rmw_implementation/)

## Testing [4]

### Feature Testing [4.i]

There are currently no public features undergoing tests.

### Public API Testing [4.ii]

`rmw_implementation` does not expose a public API. There are currently no tests for the public API.

### Coverage [4.iii]

`rmw_implementation` does not currently track test coverage.

### Performance [4.iv]

`rmw_implementation` does not currently have performance tests.

### Linters and Static Analysis [4.v]

`rmw_implementation` uses and passes all the standard linters and static analysis tools for a C++ package as described in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#linters-and-static-analysis).

Currently nightly results can be seen here:

> * [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/rmw_implementation/)

Reviewer 2: Just noticed a similar issue with your CI links here, these need to be updated here and below like the class_loader repo.

Author: I think this link is right

This document is a declaration of software quality for the `rmw_implementation` package, based on the guidelines in [REP-2004](https://www.ros.org/reps/rep-2004.html).

# `rmw_implementation` Quality Declaration

The package `rmw_implementation` claims to be in the **Quality Level 4** category.

> Below are the rationales, notes, and caveats for this claim, organized by each requirement listed in the [Package Requirements for Quality Level 1 in REP-2004](https://www.ros.org/reps/rep-2004.html).

Reviewer 2: ```suggestion
Below are the rationales, notes, and caveats for this claim, organized by each requirement listed in the [Package Requirements for Quality Level 4 in REP-2004](https://www.ros.org/reps/rep-2004.html).
```

## Version Policy [1]

### Version Scheme [1.i]

`rmw_implementation` uses `semver` according to the recommendation for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#versioning).

### Version Stability [1.ii]

`rmw_implementation` is not yet at a stable version, i.e. >= 1.0.0.

### Public API Declaration [1.iii]

> **`rmw_implementation` does not expose a public API.**

Reviewer 2: Did you mean for this to be bolded?

### API Stability Policy [1.iv]

`rmw_implementation` will not break public API within a released ROS distribution, i.e. no major releases once the ROS distribution is released.

### ABI Stability Policy [1.v]

`rmw_implementation` contains C code and therefore must be concerned with ABI stability, and will maintain ABI stability within a ROS distribution.

### ABI and ABI Stability Within a Released ROS Distribution [1.vi]

`rmw_implementation` will not break API nor ABI within a released ROS distribution, i.e. no major releases once the ROS distribution is released.
## Change Control Process [2]

`rmw_implementation` follows the recommended guidelines for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process).

### Change Requests [2.i]

This package requires that all changes occurr through a pull request.

### Contributor Origin [2.ii]

This package has a confirmation of contributor origin policy, which can be found in [CONTRIBUTING](./CONTRIBUTING.md).

### Peer Review Policy [2.iii]

> Following the recommended guidelines for ROS Core packages, all pull request have at least 1 peer review.

Reviewer: The exact number might change
```suggestion
All pull request will be peer-reviewed, check [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#change-control-process) for additional information.
```

### Continuous Integration [2.iv]

All pull request must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)

Currently nightly results can be seen here:
* [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/rmw_implementation/)

> * [linux-arm64_release](https://ci.ros2.org/view/nightly/job/nightly_linux_release/lastBuild/testReport/rmw_implementation/)

Reviewer 2: ```suggestion
* [linux_release](https://ci.ros2.org/view/nightly/job/nightly_linux_release/lastBuild/testReport/rmw_implementation/)
```

* [mac_osx_release](https://ci.ros2.org/view/nightly/job/nightly_osx_release/lastBuild/testReport/rmw_implementation/)
* [windows_release](https://ci.ros2.org/view/nightly/job/nightly_win_rel/lastBuild/testReport/rmw_implementation/)

### Documentation Policy [2.v]

> `rmw_implementation` does not expose a public API. There is no need to add documentation to new changes. In the case that this package has a public API the documetation of new functionality will be required.

Reviewer: ```suggestion
`rmw_implementation` does not expose a public API. There is no need to add documentation to new changes. In the case that this package adds public API features the documentation of new functionality will be required.
```

## Documentation [3]

### Feature Documentation [3.i]

> `rmw_implementation` does not have any feature documentation and it will need to be added for higher quality levels.

Reviewer: ```suggestion
`rmw_implementation` does not have any feature documentation and it will be needed to reach higher quality levels.
```

There is an automated test which runs a linter (ament_copyright) that ensures each file has at least one copyright statement.

Most recent test results can be found [here](http://build.ros2.org/view/Epr/job/Epr__class_loader__ubuntu_bionic_amd64/lastBuild/testReport/rmw_implementation/)

## Testing [4]

### Feature Testing [4.i]

There are currently no public features undergoing tests.

### Public API Testing [4.ii]

`rmw_implementation` does not expose a public API. There are currently no tests for the public API.

### Coverage [4.iii]

`rmw_implementation` does not currently track test coverage.

### Performance [4.iv]

`rmw_implementation` does not currently have performance tests.

### Linters and Static Analysis [4.v]

`rmw_implementation` uses and passes all the standard linters and static analysis tools for a C++ package as described in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#linters-and-static-analysis).

Currently nightly results can be seen here:
* [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/rmw_implementation/)

> * [linux-arm64_release](https://ci.ros2.org/view/nightly/job/nightly_linux_release/lastBuild/testReport/rmw_implementation/)

Reviewer 2: ```suggestion
* [linux_release](https://ci.ros2.org/view/nightly/job/nightly_linux_release/lastBuild/testReport/rmw_implementation/)
```

* [mac_osx_release](https://ci.ros2.org/view/nightly/job/nightly_osx_release/lastBuild/testReport/rmw_implementation/)
* [windows_release](https://ci.ros2.org/view/nightly/job/nightly_win_rel/lastBuild/testReport/rmw_implementation/)

## Dependencies [5]

> ### Direct Runtime ROS Dependencies [5.i/5.ii]

Reviewer: None of these dependencies is "optional" right? might be a good idea stating in 5.ii if that's the case.

`rmw_implementation` has run-time and build-time dependencies that need to be considered for this declaration.

 - rcpputils
 - rcutils
 - rmw
 - rmw_implementation_cmake

### Direct Runtime non-ROS Dependency [5.iii]

`rosidl_typesupport_cpp` does not have any runtime non-ROS dependencies.

It has one "buildtool" dependency, which do not affect the resulting quality of the package, because they do not contribute to the public library API.
It also has several test dependencies, which do not affect the resulting quality of the package, because they are only used to build and run the test code.

## Platform Support [6]

`rmw_implementation` supports all of the tier 1 platforms as described in [REP-2000](https://www.ros.org/reps/rep-2000.html#support-tiers), and tests each change against all of them.

Currently nightly results can be seen here:
* [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/rmw_implementation/)

> * [linux-arm64_release](https://ci.ros2.org/view/nightly/job/nightly_linux_release/lastBuild/testReport/rmw_implementation/)

Reviewer 2: ```suggestion
* [linux_release](https://ci.ros2.org/view/nightly/job/nightly_linux_release/lastBuild/testReport/rmw_implementation/)
```

* [mac_osx_release](https://ci.ros2.org/view/nightly/job/nightly_osx_release/lastBuild/testReport/rmw_implementation/)
* [windows_release](https://ci.ros2.org/view/nightly/job/nightly_win_rel/lastBuild/testReport/rmw_implementation/)

## Vulnerability Disclosure Policy [7.i]

This package does not yet have a Vulnerability Disclosure Policy

# Current status Summary

The chart below compares the requirements in the REP-2004 with the current state of the rcutils package.

|Number|  Requirement| Current state |
|--|--|--|
|1| **Version policy** |---|
|1.i|Version Policy available | ✓ |
|1.ii|Stable version |x|

> |1.iii|Declared public API|None|
> |1.iv|API stability policy|None|
> |1.v|ABI stability policy|None|
> |1.vi_|API/ABI stable within ros distribution|None|

Reviewer: The policies are there, although these don't apply directly to the package. I think these can be marked with a check.
```suggestion
|1.iii|Declared public API|Not required|
|1.iv|API stability policy|✓|
|1.v|ABI stability policy|✓|
|1.vi_|API/ABI stable within ros distribution|✓|
```

|2| **Change control process** |---|
|2.i| All changes occur on change request |✓|

> |2.ii| Contributor origin (DCO, CLA, etc) |x|

Reviewer: ```suggestion
|2.ii| Contributor origin (DCO, CLA, etc) |✓|
```

|2.iii| Peer review policy |✓|

> |2.iv| CI policy for change requests |x|
> |2.v| Documentation policy for change requests |☓|

Reviewer: ```suggestion
|2.iv| CI policy for change requests |✓|
|2.v| Documentation policy for change requests |✓|
```

|3| **Documentation** | --- |

> |3.i| Per feature documentation | None |
> |3.ii| Per public API item documentation | None |

Reviewer: Not required could be N/A as well
```suggestion
|3.i| Per feature documentation | ✓ |
|3.ii| Per public API item documentation | Not required |
```

|3.iii| Declared License(s) | ✓ |
|3.iv| Copyright in source files| ✓ |
|3.v.a| Quality declaration linked to README | ✓ |
|3.v.b| Centralized declaration available for peer review |✓|
|4| Testing | --- |
|4.i| Feature items tests | None |
|4.ii| Public API tests | None |
|4.iii.a| Using coverage | None |
|4.iii.a| Coverage policy | None |

> |4.iv.a| Performance tests (if applicable) | ? |

Reviewer: ```suggestion
|4.iv.a| Performance tests (if applicable) | x |
```

|4.iv.b| Performance tests policy| None |
|4.v.a| Code style enforcement (linters)| ✓ |
|4.v.b| Use of static analysis tools | None |
|5| Dependencies | --- |
|5.i| Must not have ROS lower level dependencies | ✓ |
|5.ii| Optional ROS lower level dependencies| ✓ |
|5.iii| Justifies quality use of non-ROS dependencies |✓|
|6| Platform support | --- |
|6.i| Support targets Tier1 ROS platforms| ✓ |
|7| Security | --- |

> |7.i| Vulnerability Disclosure Policy | ? |

Reviewer: ```suggestion
|7.i| Vulnerability Disclosure Policy | x |
```

> `rmw_implementation` uses `semver` according to the recommendation for ROS Core packages in the [ROS 2 Developer Guide](https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#versioning), and is at or above a stable version, i.e. `>= 1.0.0`.

Reviewer 2: >  and is at or above a stable version, i.e. `>= 1.0.0`.

Remove

### Change Requests [2.i]

This package requires that all changes occurr through a pull request.

### Contributor Origin [2.ii]

This package has a confirmation of contributor origin policy, which can be found in [CONTRIBUTING](./CONTRIBUTING.md).

### Peer Review Policy [2.iii]

Following the recommended guidelines for ROS Core packages, all pull request have at least 1 peer review.

### Continuous Integration [2.iv]

All pull request must pass CI on all [tier 1 platforms](https://www.ros.org/reps/rep-2000.html#support-tiers)

Currently nightly results can be seen here:
* [linux-aarch64_release](https://ci.ros2.org/view/nightly/job/nightly_linux-aarch64_release/lastBuild/testReport/rmw_implementation/)
* [linux-arm64_release](https://ci.ros2.org/view/nightly/job/nightly_linux_release/lastBuild/testReport/rmw_implementation/)
* [mac_osx_release](https://ci.ros2.org/view/nightly/job/nightly_osx_release/lastBuild/testReport/rmw_implementation/)
* [windows_release](https://ci.ros2.org/view/nightly/job/nightly_win_rel/lastBuild/testReport/rmw_implementation/)

### Documentation Policy [2.v]

> All pull request must resolve related documentation changes before merging

Reviewer 2: Is this true if there is no documentation?
